File: hard/123-best-time-to-buy-and-sell-stock-iii/solution.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """
        123. Best Time to Buy and Sell Stock III
        Time: O(n)
        Space: O(1)
        """
        NEG = float("-inf")
        buy1 = buy2 = NEG   # cash while holding the 1st / 2nd share
        sell1 = sell2 = 0   # cash after selling the 1st / 2nd share

        for p in prices:
            buy1 = max(buy1, -p)
            sell1 = max(sell1, buy1 + p)
            buy2 = max(buy2, sell1 - p)
            sell2 = max(sell2, buy2 + p)

        return sell2

File: hard/123-best-time-to-buy-and-sell-stock-iii/solution.java
/**
 * 123. Best Time to Buy and Sell Stock III
 * Time: O(n)
 * Space: O(1)
 */
class Solution {
    public int maxProfit(int[] prices) {
        // cash while holding the 1st / 2nd share, and after selling each
        int buy1 = Integer.MIN_VALUE, buy2 = Integer.MIN_VALUE;
        int sell1 = 0, sell2 = 0;

        for (int p : prices) {
            buy1 = Math.max(buy1, -p);
            sell1 = Math.max(sell1, buy1 + p);
            buy2 = Math.max(buy2, sell1 - p);
            sell2 = Math.max(sell2, buy2 + p);
        }

        return sell2;
    }
}
